### **Enunciado del Ejercicio Práctico - Núcleo Formativo 1**

**Título del Ejercicio:** Creación del Repositorio Unificado de Actividad del Cliente

**Contexto y Escenario:**
Eres el/la primer/a Ingeniero/a de Datos en "ElectroShop", una prometedora tienda online de electrónica. Las ventas están creciendo exponencialmente, pero también lo está haciendo un problema silencioso y costoso: las **transacciones fraudulentas**. La empresa está perdiendo miles de euros por compras realizadas con tarjetas de crédito robadas y por actividades de bots.

El equipo de prevención de fraude está desbordado. La información que necesitan está dispersa en tres sistemas críticos que no se comunican entre sí:
*   La **base de datos de transacciones** con los registros de pagos.
*   Los **logs del servidor web**, que registran cada clic y página vista por los usuarios.
*   El **sistema de ticketing**, donde los clientes reales reportan problemas (a menudo, cargos fraudulentos en sus cuentas).

**Tu Misión:**
Tu misión es construir la base de todo el futuro sistema de detección de fraude: un **Data Lake seguro y centralizado** que integre estas tres fuentes de datos. Debes capturar y estructurar esta información masiva para permitir que los analistas y científicos de datos puedan, en los siguientes módulos, construir modelos para detectar anomalías en los flujos de compra.

**Objetivos de Aprendizaje:**
*   Aplicar técnicas de ingesta de datos desde formatos variados (CSV, JSON, y logs semi-estructurados).
*   Desarrollar un proceso sistemático de diagnóstico y limpieza de datos, crucial para la precisión de cualquier algoritmo de fraude.
*   Resolver problemas comunes: valores nulos en transacciones, parsing de logs complejos y enriquecimiento de datos a partir de diferentes fuentes.
*   Enriquecer los datos mediante Feature Engineering para crear señales útiles para la detección de fraude.
*   Seleccionar y aplicar un formato de almacenamiento (Parquet) que sea eficiente para análisis de seguridad.
*   Justificar las decisiones técnicas tomadas.

**Ficheros de Datos Proporcionados (Generados por un script):**
1.  `transacciones.csv`: Un registro de todas las transacciones de pago, con valores nulos y estados inconsistentes.
2.  `tickets_soporte.json`: Tickets de soporte en formato JSON, incluyendo reportes de fraude por parte de clientes.
3.  `logs_web.txt`: Un fichero de logs semi-estructurado con la actividad de navegación de los usuarios, un formato muy común pero difícil de parsear.

**Tareas a Realizar en tu Jupyter Notebook:**

1.  **Fase 1: Configuración e Ingesta de Datos**
    *   Instala las librerías necesarias.
    *   Carga `transacciones.csv` y `tickets_soporte.json` en DataFrames de Pandas.
    *   **Reto:** Carga el fichero `logs_web.txt` línea por línea y utiliza **expresiones regulares (regex)** o métodos de `split` para parsearlo y convertirlo en un DataFrame estructurado con columnas como `timestamp`, `ip_address`, `user_id`, `http_method`, `url`, etc.

2.  **Fase 2: Diagnóstico y Análisis Exploratorio**
    *   Para cada DataFrame, realiza un diagnóstico para identificar problemas de calidad, como tipos de datos incorrectos, valores nulos, etc.

3.  **Fase 3: Limpieza y Transformación**
    *   **Para `df_transacciones`:**
        *   Asegúrate de que `fecha_transaccion` sea del tipo `datetime`.
        *   Estandariza la columna `estado` (ej. "approved", "APPROVED", "success" -> "APROBADA").
        *   Gestiona los valores nulos en `metodo_pago`.
    *   **Para `df_tickets`:**
        *   Normaliza la columna anidada `detalles_ticket`.
        *   Crea una columna booleana `es_reporte_fraude` basada en el texto del ticket.
    *   **Para `df_logs`:**
        *   Limpia y convierte los tipos de datos de las columnas parseadas (ej. `user_id` a numérico, `status_code` a entero).

4.  **Fase 4: Enriquecimiento de Datos (Feature Engineering)**
    *   **Unión de Datos:** Une (`merge`) el DataFrame de transacciones con el de tickets para identificar qué transacciones fueron posteriormente reportadas como fraude.
    *   Crea una nueva columna `hora_transaccion` en el DataFrame de transacciones.

5.  **Fase 5: Almacenamiento Optimizado en el Data Lake**
    *   Crea un directorio llamado `ecommerce_data_lake/raw` y otro `ecommerce_data_lake/processed`.
    *   Guarda los DataFrames limpios y enriquecidos en `ecommerce_data_lake/processed` en formato **Parquet**.

6.  **Fase 6: Conclusiones y Justificación**
    *   En una celda de Markdown, responde a:
        *   ¿Por qué fue el parsing de los logs web (`logs_web.txt`) el paso de ingesta más complejo?
        *   ¿Qué ventaja clave ofrece tener un Data Lake unificado con transacciones, logs y tickets para la detección de fraude, en comparación con tenerlos en sistemas separados?

**Entregable:**
*   <span style="color: red;">Un único fichero Jupyter Notebook (`.ipynb`) con todo el código, las salidas y las conclusiones.</span>

---

### **Rúbrica de Evaluación para el Ejercicio Práctico (NF1)**

| Criterio | Insuficiente (0-4 pts) | Suficiente (5-6 pts) | Notable (7-8 pts) | Excelente (9-10 pts) | Puntuación |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **1. Ingesta y Parseo (20%)** | No carga todos los ficheros o el parseo de logs falla completamente. | Carga CSV y JSON, pero el parseo de logs es muy incompleto o incorrecto. | Carga todos los ficheros y logra parsear los logs en un DataFrame, aunque puede perder alguna columna. | Carga todos los ficheros de forma robusta y realiza un parseo preciso de los logs web (ej. con regex), extrayendo todas las columnas clave. | **/20** |
| **2. Limpieza de Transacciones (15%)** | No logra limpiar el DataFrame o introduce más errores. | Limpia una de las dos tareas (fechas o estados) pero no ambas, o lo hace con errores. | Limpia y estandariza correctamente las fechas y los estados de las transacciones. | Resuelve perfectamente la limpieza de fechas, la estandarización de categorías y la gestión de nulos. El código es limpio y eficiente. | **/15** |
| **3. Limpieza de Tickets y Logs (15%)** | No realiza la limpieza o el resultado es incorrecto. | Realiza la limpieza de uno de los dos DataFrames, pero falla en el otro. | Normaliza el JSON de tickets y limpia los tipos de datos de los logs de forma funcional. | Normaliza el JSON, crea la columna `es_reporte_fraude` de forma precisa, y asegura que todas las columnas de los logs tengan los tipos de datos correctos. | **/15** |
| **4. Feature Engineering (15%)** | No crea ninguna columna nueva ni une los DataFrames. | Intenta unir los DataFrames pero la lógica es incorrecta, o crea una característica simple pero no realiza la unión. | Une correctamente transacciones y tickets, identificando las transacciones fraudulentas reportadas. | Realiza la unión de forma impecable y crea al menos una nueva característica útil adicional (ej. `hora_transaccion`), demostrando una buena comprensión de la preparación de datos. | **/15** |
| **5. Almacenamiento de Datos (15%)** | No guarda los ficheros o lo hace en un formato incorrecto. | Guarda los ficheros en Parquet, pero no en la estructura de directorios solicitada. | Guarda los ficheros en formato Parquet y en la estructura de directorios correcta. | Guarda perfectamente los DataFrames limpios y enriquecidos en formato Parquet dentro de la estructura `processed`, demostrando entender la organización de un Data Lake. | **/15** |
| **6. Calidad y Justificación (20%)** | El notebook es desordenado, sin comentarios y sin conclusiones. | El código funciona pero es difícil de leer. La justificación es muy breve o incorrecta. | El notebook está bien estructurado. La justificación técnica es correcta pero superficial. | El notebook es ejemplar: claro, bien comentado y con celdas de Markdown. La justificación demuestra una profunda comprensión del valor de negocio de unificar los datos. | **/20** |
| **TOTAL** | | | | | **/100** |

# Fase 1: Configuración e Ingesta de Datos

In [1]:
# Instalar e importar las librerías necesarias

!pip install pandas seaborn matplotlib numpy

import os
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from datetime import datetime

In [2]:
# Cargar transacciones.csv y tickets_soporte.json en DataFrames de Pandas

df_transacciones = pd.read_csv("transacciones.csv") 
df_tickets = pd.read_json("tickets_soporte.json")

print("--- Fichero correspondiente a las transacciones realizadas ---\n")
print(df_transacciones.head(), "\n")
print("--- Fichero correspondiente a los tickets de soporte ---\n")
print(df_tickets.head(), "\n")

# Cargar logs_web.txt

registros = []

with open("logs_web.txt", "r", encoding="utf-8") as f:
    for linea in f:
        registros.append([registro.strip() for registro in linea.split("--")]) # Utilizar "--" como separador entre datos dentro de cada línea de registro

df_logs = pd.DataFrame(registros, columns=[
    "timestamp",
    "ip_address",
    "user_id",
    "connection",
    "status_code"
])

# Utilizar expresiones regulares (RegEx) para terminar de obtener todos los datos de cada registro de forma eficiente

df_logs["user_id"] = df_logs["user_id"].str.replace("user:", "", regex=False)

regex = r'"(\w+)"\s+(\S+)\s+(HTTP/\d\.\d)'
df_logs[["method", "endpoint", "http_version"]] = df_logs["connection"].str.extract(regex)
df_logs.drop(columns=["connection"], inplace=True)

print("--- Fichero correspondiente a logs en la web ---\n")
print(df_logs.head(20), "\n")

--- Fichero correspondiente a las transacciones realizadas ---

                         id_transaccion  user_id           fecha_transaccion  \
0  d1d7c920-906f-45fe-b8b6-c32e95f3ef5b     1563  2025-02-08T09:56:35.425203   
1  cb0c0177-32bc-42ac-a03f-46054b93632f     1213  2025-06-11T06:26:45.131891   
2  d935437d-43c9-47bd-8a8e-34610cd81bd2     1033  2025-04-18T12:01:43.479937   
3  cb7fb6e8-92d6-45ce-b8b9-1fdd06855655     1464  2025-06-08T05:16:24.747770   
4  977163ae-de58-4664-90e7-5cf5658bca2f     1089  2025-02-26T19:01:14.961432   

     monto  metodo_pago    estado  
0  1191.40  Credit Card   pending  
1   616.34  Credit Card   pending  
2   356.91       PayPal   success  
3    90.33          NaN   success  
4   678.56          NaN  APPROVED   

--- Fichero correspondiente a los tickets de soporte ---

   id_ticket              fecha_creacion  \
0      91681  2025-06-14T19:45:31.338767   
1      90813  2025-08-03T19:45:31.325507   
2      90328  2025-08-04T19:45:31.323935   
3  

# Fase 2: Diagnóstico y Análisis Exploratorio

In [3]:
# Para cada DataFrame, realiza un diagnóstico para identificar problemas de calidad, como tipos de datos incorrectos, valores nulos, etc.

def diagnostico_calidad(df, nombre_df):
    print(f"\n{'='*60}")
    print(f"DIAGNÓSTICO DEL DATAFRAME: {nombre_df}")
    print(f"{'='*60}")
    
    # Información general
    print("\n--- INFO GENERAL ---")
    df.info()
    
    # Conteo de nulos reales
    print("\n--- NULOS REALES (isnull) ---")
    print(df.isnull().sum())
    
    print("\n--- % DE NULOS ---")
    print((df.isnull().mean() * 100).round(2))
    
    # Valores sospechosos típicos
    print("\n--- VALORES SOSPECHOSOS ---")
    valores_sospechosos = ["-", "null", "NULL", "None", "", " "]
    
    for val in valores_sospechosos:
        conteo = (df == val).sum().sum()
        if conteo > 0:
            print(f"Valor '{val}' encontrado {conteo} veces")
    
    # Detección de tipos incorrectos
    print("\n--- COLUMNAS OBJECT (posibles tipos incorrectos) ---")
    print(df.select_dtypes(include=["object", "string"]).columns.tolist())

diagnostico_calidad(df_transacciones, "Transacciones")
diagnostico_calidad(df_tickets, "Tickets Soporte")
diagnostico_calidad(df_logs, "Logs Web")


DIAGNÓSTICO DEL DATAFRAME: Transacciones

--- INFO GENERAL ---
<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id_transaccion     50000 non-null  str    
 1   user_id            50000 non-null  int64  
 2   fecha_transaccion  50000 non-null  str    
 3   monto              50000 non-null  float64
 4   metodo_pago        37664 non-null  str    
 5   estado             50000 non-null  str    
dtypes: float64(1), int64(1), str(4)
memory usage: 5.9 MB

--- NULOS REALES (isnull) ---
id_transaccion           0
user_id                  0
fecha_transaccion        0
monto                    0
metodo_pago          12336
estado                   0
dtype: int64

--- % DE NULOS ---
id_transaccion        0.00
user_id               0.00
fecha_transaccion     0.00
monto                 0.00
metodo_pago          24.67
estado                0.00
dtype: fl

## 📊 Conclusión del Diagnóstico Inicial

Tras analizar los tres DataFrames, se observan los siguientes puntos clave:

### Transacciones
- 50.000 registros.
- La columna `metodo_pago` tiene un 24,67% de valores nulos.
- `fecha_transaccion` está como `str` y debería convertirse a `datetime`.
- El resto de columnas no presenta problemas aparentes.

### Tickets Soporte
- 1.762 registros.
- No hay valores nulos.
- `fecha_creacion` también debería convertirse a `datetime`.
- `detalles_ticket` es texto libre (posible análisis posterior).

### Logs Web
- 2.000.000 registros (dataset grande).
- No hay nulos reales, pero existen 599.946 valores `"-"` que probablemente representan datos faltantes.
- Todas las columnas están como `str` y requieren tipado adecuado (`datetime`, `int`, etc.).

---

## 🎯 Resumen General

El principal trabajo de limpieza consistirá en:

- Normalizar valores faltantes (`"-"` → `NaN`).
- Convertir columnas de fechas a `datetime`.
- Ajustar tipos de datos para mejorar calidad y eficiencia.
- Revisar la columna `metodo_pago` por su alto porcentaje de nulos.

Con esto, los DataFrames quedarán listos para la fase de transformación y feature engineering dentro del pipeline ETL.

In [4]:
# Antes de empezar con la fase de limpieza y transformación de datos, se procede a convertir los valores catalogados como sospechosos en valores de tipo NaN

# Valores sospechosos típicos
valores_nulos = ["-", "null", "NULL", "None", "", " "]

# Reemplazar en cada DataFrame
df_transacciones.replace(valores_nulos, np.nan, inplace=True)
df_tickets.replace(valores_nulos, np.nan, inplace=True)
df_logs.replace(valores_nulos, np.nan, inplace=True)

print("Valores nulos normalizados a NaN")

print(df_transacciones.isna().sum())
print(df_tickets.isna().sum())
print(df_logs.isna().sum())

Valores nulos normalizados a NaN
id_transaccion           0
user_id                  0
fecha_transaccion        0
monto                    0
metodo_pago          12336
estado                   0
dtype: int64
id_ticket          0
fecha_creacion     0
detalles_ticket    0
dtype: int64
timestamp            0
ip_address           0
user_id         599946
status_code          0
method               0
endpoint             0
http_version         0
dtype: int64


# Fase 3: Limpieza y Transformación

In [5]:
## Para df_transacciones:
## - Asegúrate de que fecha_transaccion sea del tipo datetime.
## - Estandariza la columna estado (ej. "approved", "APPROVED", "success" -> "APROBADA").
## - Gestiona los valores nulos en metodo_pago.

print("\n--- Tipo de dato de la columna 'fecha_transaccion' del DataFrame de transacciones ---\n")
df_transacciones["fecha_transaccion"] = pd.to_datetime(df_transacciones["fecha_transaccion"], errors="coerce")
print(df_transacciones["fecha_transaccion"].dtype)
print(f"Nulos en la columna: {df_transacciones['fecha_transaccion'].isna().sum()}")

print("\n--- Valores existentes en la columna 'estado' del DataFrame de transacciones ---\n")
df_transacciones.loc[df_transacciones["estado"].isin(["approved", "APPROVED", "success"]), "estado"] = "APROBADA"
df_transacciones.loc[df_transacciones["estado"].isin(["denied", "failed"]), "estado"] = "RECHAZADA"
df_transacciones.loc[df_transacciones["estado"].isin(["pending"]), "estado"] = "PENDIENTE"
print(df_transacciones["estado"].value_counts(dropna=False))

print("\n--- Normalización de valores nulos en la columna 'metodo_pago' del DataFrame de transacciones ---\n")
df_transacciones["metodo_pago"] = df_transacciones["metodo_pago"].str.upper()
df_transacciones["metodo_pago"] = df_transacciones["metodo_pago"].fillna("OTHER")
print(df_transacciones["metodo_pago"].isna().sum())
print(df_transacciones["metodo_pago"].value_counts(dropna=False).head(10))

## Para df_tickets:
## - Normaliza la columna anidada detalles_ticket.
## - Crea una columna booleana es_reporte_fraude basada en el texto del ticket.

print("\n--- Conversión a tipo Datetime de la columna 'fecha_creacion' en el DataFrame de tickets ---\n")
df_tickets["fecha_creacion"] = pd.to_datetime(df_tickets["fecha_creacion"], errors="coerce")

print("\n--- Normalización de la columna anidada 'detalles_ticket' del DataFrame de tickets ---\n")
detalles_json = pd.json_normalize(df_tickets["detalles_ticket"])
df_tickets = pd.concat([df_tickets, detalles_json], axis=1)
df_tickets.drop(columns=["detalles_ticket"], inplace=True)
df_tickets.info()

print("\n--- Crear columna booleana 'es_reporte_fraude' del DataFrame de tickets ---\n")
palabras_fraude = ["fraude", "fraud", "robo", "hack", "phishing"]
df_tickets["descripcion"] = df_tickets["descripcion"].str.lower()
df_tickets["es_reporte_fraude"] = False
df_tickets.loc[
    df_tickets["descripcion"].str.contains(
        "|".join(palabras_fraude),
        case=False,
        na=False
    ), 
    "es_reporte_fraude"    
] = True
print(df_tickets["es_reporte_fraude"].value_counts(normalize=True, dropna=False) * 100)

## Para df_logs:
## - Limpia y convierte los tipos de datos de las columnas parseadas (ej. user_id a numérico, status_code a entero).

print("\n--- Convertimos los tipos de datos correctamente del DataFrame de logs ---\n")

df_logs["timestamp"] = pd.to_datetime(df_logs["timestamp"], errors="coerce")
df_logs["user_id"] = pd.to_numeric(df_logs["user_id"], errors="coerce")
df_logs["status_code"] = pd.to_numeric(df_logs["status_code"], errors="coerce").astype('Int64')

df_logs.info() # Verificamos los tipos de datos que tenemos ahora


--- Tipo de dato de la columna 'fecha_transaccion' del DataFrame de transacciones ---

datetime64[us]
Nulos en la columna: 0

--- Valores existentes en la columna 'estado' del DataFrame de transacciones ---

estado
APROBADA     26322
RECHAZADA    15638
PENDIENTE     8040
Name: count, dtype: int64

--- Normalización de valores nulos en la columna 'metodo_pago' del DataFrame de transacciones ---

0
metodo_pago
CREDIT CARD    12651
DEBIT CARD     12592
PAYPAL         12421
OTHER          12336
Name: count, dtype: int64

--- Conversión a tipo Datetime de la columna 'fecha_creacion' en el DataFrame de tickets ---


--- Normalización de la columna anidada 'detalles_ticket' del DataFrame de tickets ---

<class 'pandas.DataFrame'>
RangeIndex: 1762 entries, 0 to 1761
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id_ticket       1762 non-null   int64         
 1   fecha_creacion  1762 non-null   datet

Aunque la tasa de **es_reporte_fraude = True (71.6%)** es correcta desde el punto de vista matemático (1262 de 1762 tickets), este resultado requiere un análisis más profundo debido a su magnitud.

En primer lugar, se trata de una proporción anormalmente elevada si se compara con escenarios reales, donde los casos de fraude suelen representar un porcentaje mucho menor del total. Esto sugiere que el dataset podría no ser representativo de la realidad general, sino estar condicionado por su origen. Por ejemplo, es posible que los datos provengan de un sistema especializado en la gestión de incidencias relacionadas con fraude, lo que explicaría la sobrerrepresentación de este tipo de casos.

Por otro lado, también cabe considerar la posibilidad de que exista algún tipo de sesgo en la recogida o en el etiquetado de los datos, ya sea por criterios amplios a la hora de clasificar un ticket como fraude o por una selección previa de los casos incluidos en el dataset.

Desde el punto de vista analítico, este desequilibrio tiene implicaciones importantes. En caso de utilizar estos datos para entrenar modelos de machine learning, la alta proporción de una clase frente a la otra podría provocar que el modelo aprenda patrones sesgados, favoreciendo la clase mayoritaria. Esto haría necesario aplicar técnicas de balanceo o utilizar métricas de evaluación adecuadas como precision, recall o F1-score, en lugar de basarse únicamente en la exactitud (accuracy).

En conclusión, aunque el cálculo es correcto, el valor obtenido debe ser interpretado con cautela, analizando el contexto y las posibles limitaciones del dataset.

# Fase 4: Enriquecimiento de Datos (Feature Engineering)

In [6]:
## Unión de Datos: Une (merge) el DataFrame de transacciones con el de tickets para identificar qué transacciones fueron posteriormente reportadas como fraude.
## Crea una nueva columna hora_transaccion en el DataFrame de transacciones.

print(df_tickets["descripcion"].head(20))

# Observamos que en algunos tickets aparece el número de transacción, así que siguiente paso: 
# Crear la columna id_transaccion en df_tickets a partir de la descripcion para poder mergear ambos DataFrames

df_tickets["id_transaccion"] = df_tickets["descripcion"].str.extract(
    r'([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})',
    expand=False
)
df_tickets["id_transaccion"] = df_tickets["id_transaccion"].fillna("SIN_TRANSACCION")

print(f"\nValores de la nueva columna id_transaccion:\n {df_tickets['id_transaccion'].value_counts(dropna=False).head(10)}\n")

df_transacciones_tickets = pd.merge(
    df_transacciones,
    df_tickets,
    on="id_transaccion",
    how="left"
)

print("\n--- Nuevo DataFrame resultante de mergear transacciones y tickets ---\n")

df_transacciones_tickets.info()

print("\n--- Crear nueva columna hora_transaccion en el DataFrame de transacciones ---\n")
df_transacciones["hora_transaccion"] = df_transacciones["fecha_transaccion"].dt.hour
print(f"Valores de hora_transaccion: {df_transacciones['hora_transaccion'].value_counts(dropna=False).head(10)}")

0        despite present play hair crime stand century.
1     he visto un cargo asociado a la transacción cd...
2     he visto un cargo asociado a la transacción 41...
3     small growth economy his pretty forget financi...
4     he visto un cargo asociado a la transacción ee...
5     he visto un cargo asociado a la transacción 39...
6                     option mind town evidence kid no.
7     he visto un cargo asociado a la transacción 71...
8                           among prepare ball herself.
9     he visto un cargo asociado a la transacción c8...
10    he visto un cargo asociado a la transacción a4...
11    he visto un cargo asociado a la transacción ae...
12    he visto un cargo asociado a la transacción 7a...
13    he visto un cargo asociado a la transacción 90...
14    he visto un cargo asociado a la transacción 7a...
15    he visto un cargo asociado a la transacción 71...
16    he visto un cargo asociado a la transacción 38...
17    he visto un cargo asociado a la transacció

# Fase 5: Almacenamiento Optimizado en el Data Lake

In [7]:
# Crear estructura Data Lake
os.makedirs("ecommerce_data_lake/raw", exist_ok=True)
os.makedirs("ecommerce_data_lake/processed", exist_ok=True)

# Transacciones
df_transacciones.to_parquet(
    "ecommerce_data_lake/processed/transacciones.parquet",
    index=False
)

# Logs
df_logs.to_parquet(
    "ecommerce_data_lake/processed/logs.parquet",
    index=False
)

# Tickets
df_tickets.to_parquet(
    "ecommerce_data_lake/processed/tickets.parquet",
    index=False
)

# Transacciones + Tickets
df_transacciones_tickets.to_parquet(
    "ecommerce_data_lake/processed/transacciones_tickets.parquet",
    index=False
)

print("Data Lake creado y procesado correctamente.")

Data Lake creado y procesado correctamente.


# Fase 6: Conclusiones y Justificación

### ¿Por qué fue el parsing de los logs web (logs_web.txt) el paso de ingesta más complejo?

El parsing de los logs web (logs_web.txt) representó el paso de ingesta más complejo debido a que se trataba de datos semi-estructurados, a diferencia de otros formatos más estructurados como CSV o JSON. Los logs web contenían múltiples atributos codificados dentro de una misma línea de texto, por lo que fue necesario aplicar técnicas de procesamiento de texto y extracción de patrones para transformar la información en un formato tabular adecuado para su análisis mediante Pandas.

En concreto, fue necesario identificar y separar campos como timestamp, dirección IP, identificador de usuario, método HTTP, endpoint, versión del protocolo y código de respuesta utilizando delimitadores y expresiones regulares. Este proceso implica una mayor complejidad lógica en comparación con la carga directa de datasets estructurados, ya que el parsing debe ser capaz de manejar variaciones en el formato del texto y posibles inconsistencias en los registros. De hecho, esta dificultad se volvió a ver a la hora de extraer el ID de transacción desde la descripción de los diferetes tickets.

Adicionalmente, fue necesario realizar un proceso de limpieza posterior para tratar representaciones no estándar de datos, como caracteres especiales o valores simbólicos como -, que debieron ser convertidos a valores nulos para mantener la consistencia del dataset.

Desde el punto de vista del pipeline de datos, este paso es fundamental en arquitecturas ETL modernas, ya que garantiza que los datos puedan ser posteriormente utilizados en tareas de análisis exploratorio o modelado predictivo.

### ¿Qué ventaja clave ofrece tener un Data Lake unificado con transacciones, logs y tickets para la detección de fraude, en comparación con tenerlos en sistemas separados?

La principal ventaja de un Data Lake unificado radica en la capacidad de realizar un análisis exploratorio entre múltiples fuentes de datos, permitiendo obtener una visión general del comportamiento del usuario y mejorar la detección de actividades fraudulentas. En sistemas tradicionales, la información de transacciones, actividad web y tickets de soporte se encuentra distribuida en diferentes sistemas, lo que limita la capacidad de detectar patrones complejos de fraude que requieren la combinación de múltiples señales de comportamiento.

En cambio, un Data Lake permite integrar estos datos en un mismo entorno, facilitando la construcción de features más ricas para modelos de ML y análisis de comportamiento. Por ejemplo, es posible correlacionar:

* Eventos de navegación detectados en los logs web.
* Operaciones financieras registradas en las transacciones.
* Reportes de incidencias o fraude detectados en tickets de soporte.

Desde la perspectiva de la Inteligencia Artificial y el Análisis de Datos, esta unificación permite mejorar la capacidad predictiva de los modelos al proporcionar contexto adicional, lo que resulta especialmente útil en sistemas de detección de fraude basados en aprendizaje supervisado o análisis de anomalías. Desde el punto de vista del negocio, un Data Lake unificado mejora la toma de decisiones basada en datos y permite la implementación de sistemas de monitorización y alerta temprana frente a comportamientos sospechosos.

# Conclusión

La construcción de un pipeline de datos que incluya ingesta, limpieza, transformación y unificación en un Data Lake constituye la base para el desarrollo de sistemas analíticos avanzados y aplicaciones de IA, especialmente en escenarios de detección de fraude y análisis del comportamiento del usuario.